shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,183 @@ use crate::{
 pub struct BalancerPools {
     /// Used for O(1) access to all pool_ids for a given token
     pools_by_token: HashMap<H160, HashSet<H256>>,
     /// WeightedPool data for a given PoolId
     pools: HashMap<H256, WeightedPool>,
-    /// Temporary storage for WeightedPools containing insufficient constructor data
-    pending_pools: HashMap<H256, WeightedPoolBuilder>,
-}
-
-/// There are three specialization settings for Pools,
-/// which allow for cheaper swaps at the cost of reduced functionality:
-#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
-#[repr(u8)]
-pub enum PoolSpecialization {
-    /// no specialization, suited for all Pools. IGeneralPool is used for swap request callbacks,
-    /// passing the balance of all tokens in the Pool. These Pools have the largest swap costs
-    /// (because of the extra storage reads), which increase with the number of registered tokens.
-    General = 0,
-    /// IMinimalSwapInfoPool is used instead of IGeneralPool, which saves gas by only passing the
-    /// balance of the two tokens involved in the swap. This is suitable for some pricing algorithms,
-    /// like the weighted constant product one popularized by Balancer V1. Swap costs are
-    /// smaller compared to general Pools, and are independent of the number of registered tokens.
-    MinimalSwapInfo = 1,
-    /// only allows two tokens to be registered. This achieves the lowest possible swap gas cost.
-    /// Like minimal swap info Pools, these are called via IMinimalSwapInfoPool.
-    TwoToken = 2,
-}
-
-impl PoolSpecialization {
-    fn new(specialization: u8) -> Result<Self> {
-        match specialization {
-            0 => Ok(Self::General),
-            1 => Ok(Self::MinimalSwapInfo),
-            2 => Ok(Self::TwoToken),
-            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
-        }
-    }
+    /// Temporary storage for WeightedPools having errored on data fetching
+    pending_pools: HashSet<WeightedPoolBuilder>,
+    #[derivative(Debug = "ignore")]
+    data_fetcher: Box<dyn PoolDataFetching>,
 }
 
 impl BalancerPools {
-    fn try_upgrade(&mut self) {
-        for (pool_id, pool_builder) in self.pending_pools.clone() {
-            if let Some(weighted_pool) = pool_builder.into_pool() {
-                // When upgradable, delete pending pool and add to valid pools
-                tracing::info!("Upgrading Pool Builder with id {:?}", pool_id);
-                self.pools.insert(pool_id, weighted_pool.clone());
-                self.pending_pools.remove(&pool_id);
-                for token in weighted_pool.tokens {
-                    self.pools_by_token
-                        .entry(token)
-                        .or_default()
-                        .insert(pool_id);
-                }
+    async fn try_upgrade(&mut self) -> Result<()> {
+        for pool_builder in self.pending_pools.clone() {
+            let weighted_pool = pool_builder.into_pool(&*self.data_fetcher).await?;
```

Comment:
This part of the EVM calls could probably be parallelized. Although I would prefer to do this in a separate PR.